### ISSUE TITLE
Nullary private detection is too greedy

### ISSUE THREAD

**Issue by paracycle** [sorbet/sorbet#3872]
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0A%0Aclass%20Foo%0A%20%20def%20self.foo(%26blk)%3B%20end%0A%0A%20%20foo%20do%0A%20%20%20%20private%0A%0A%20%20%20%20def%20bar%3B%20end%0A%20%20end%0A%0A%20%20def%20baz%3B%20end%0Aend%0A%0AFoo.new.baz)

```ruby
# typed: true

class Foo
  def self.foo(&blk); end

  foo do
    private

    def bar; end
  end

  def baz; end
end

Foo.new.baz # Non-private call to private method baz on Foo https://srb.help/7031
```

#### Observed output

```
editor.rb:15: Non-private call to private method baz on Foo https://srb.help/7031
    15 |Foo.new.baz
        ^^^^^^^^^^^
    editor.rb:12: Defined in Foo here
    12 |  def baz; end
          ^^^^^^^
Errors: 1
```

#### Expected behavior

No errors since `baz` is a public method of `Foo`

- - -

The nullary `private` call in the block passed to `Foo.foo` method only modifies the visibility of the methods defined in the block itself. However, Sorbet leaks the visibility change to the scope below it as well. I suspect a similar situation happens with `protected` too.

**Comment by elliottt** [comment]
Thanks for finding this! I think that we need to be pushing/popping entries from the visibility stack for method definitions as well, in the namer

**Comment by paracycle** [comment]
Fixed by #3898
